packages/embed-react/src/contexts/token-context.tsx

```diff
@@ -7,36 +7,114 @@ import {
 import { type WealthSweetToken } from "../utils/token-utils";
 import { createContextAndHook } from "./create-context-and-hook";
 
+/**
+ * Represents an error that occurred during token generation.
+ */
+export type TokenError = {
+  message: "Failed to generate token";
+  error: unknown;
+};
+
+/**
+ * Represents the current state of token fetching.
+ */
+export type TokenFetchState = "INITIALISED" | "FETCHING" | "FETCHED" | "ERROR";
+
+/**
+ * Creates a context and hooks for managing token state.
+ */
 export const [TokenContext, useTokenContext, useTokenContextWithoutGuarantee] =
-  createContextAndHook<Partial<WealthSweetToken>>("TokenContext");
+  createContextAndHook<
+    | {
+        _tag: "success";
+        tokenFetchState: TokenFetchState;
+        token?: WealthSweetToken;
+      }
+    | { _tag: "error"; tokenFetchState: TokenFetchState; error: TokenError }
+  >("TokenContext");
 
+/**
+ * Constant representing one minute in milliseconds.
+ */
 const ONE_MINUTE = 1000 * 60;
 
+/**
+ * Props for the TokenProvider component.
+ */
+export type TokenProviderProps = PropsWithChildren<{
+  /** Function to fetch a new token */
+  fetchToken: () => Promise<WealthSweetToken>;
+  /** Optional callback for handling token fetch errors */
+  onFetchTokenError?: (error: TokenError) => void;
+}>;
+
+/**
+ * Provider component for managing token state and fetching. Handles automatic token refresh
+ * and error handling for token generation.
+ * @param props - The component props
+ * @param props.children - Child components that will have access to the token context
+ * @param props.fetchToken - Function that returns a Promise resolving to a new WealthSweetToken
+ * @param props.onFetchTokenError - Optional callback that will be called if token fetching fails
+ */
 export function TokenProvider({
   children,
   fetchToken,
-}: PropsWithChildren<{ fetchToken: () => Promise<WealthSweetToken> }>) {
+  onFetchTokenError,
+}: TokenProviderProps) {
   const [token, setToken] = useState<WealthSweetToken>();
+  const [tokenFetchState, setTokenFetchState] =
+    useState<TokenFetchState>("INITIALISED");
+  const [error, setError] = useState<TokenError>();
+
+  /**
+   * Handles token fetch errors.
+   * @param error - The error that occurred during token fetching
+   */
+  const handleTokenError = useCallback(
+    (error: unknown) => {
+      const tokenErrror = {
+        message: "Failed to generate token",
+        error,
+      } as const;
+      setError(tokenErrror);
+      setTokenFetchState("ERROR");
+      if (onFetchTokenError) {
+        onFetchTokenError(tokenErrror);
+      }
+    },
+    [setError, onFetchTokenError, setTokenFetchState],
+  );
 
+  /**
+   * Generates a new token and schedules the next token generation.
+   */
   const generateToken = useCallback(async () => {
```

Comment:
I'm curious if I ran the code

```
await generateToken();
await generateToken();
```

Would that schedule two timers which would go off later on?

Reply:
Yep I believe so. This function pushes the timeout function to the event queue every time it is called. Calling it twice would have them both added to the queue at the same time and would run in parallel (since their timouts would be essentially the same)

I dont think that is as issue since the only time we should be calling this is when generate token changes (in the useEffect)